File: q_branch/fine-grained-monitor/scripts/oscillation_detector.py
#!/usr/bin/env -S uv run
# /// script
# requires-python = ">=3.10"
# dependencies = [
#     "pyarrow>=15.0.0",
#     "pandas>=2.0.0",
#     "numpy>=1.24.0",
#     "plotly>=5.0.0",
#     "kaleido>=0.2.1",
# ]
# ///
"""
Detect CPU oscillation patterns in container metrics using autocorrelation.

This is a standalone implementation of the algorithm from the cpu_oscillation
check in the Datadog Agent (sopell/q-branch-cpu-oscillation-per-container branch).

The algorithm detects periodic CPU usage patterns by:
1. Computing CPU usage deltas from cgroup counter metrics
2. Running autocorrelation analysis to find periodicity
3. Detecting oscillation when both periodicity score and amplitude exceed thresholds

Key parameters (matching Go implementation):
- window_size: 60 samples (60 seconds at 1Hz)
- min_periodicity_score: 0.6 (autocorrelation threshold)
- min_amplitude: 10.0 (% CPU swing threshold)
- min_period: 2 (minimum period in seconds, Nyquist limit)
- max_period: 30 (maximum period in seconds)

Usage:
    uv run scripts/oscillation_detector.py metrics.parquet
    uv run scripts/oscillation_detector.py metrics.parquet --plot
    uv run scripts/oscillation_detector.py metrics.parquet --output results.html
    uv run scripts/oscillation_detector.py metrics.parquet --threshold 0.5 --min-amplitude 5.0

Reference:
    pkg/collector/corechecks/containers/cpu_oscillation/detector.go
"""

import argparse
import sys
from dataclasses import dataclass, field
from pathlib import Path

import numpy as np
import pandas as pd
import plotly.graph_objects as go
import pyarrow.parquet as pq
from plotly.subplots import make_subplots


@dataclass
class OscillationConfig:
    """Configuration for oscillation detection (mirrors Go OscillationConfig)."""

    window_size: int = 60  # Number of samples in analysis window
    min_periodicity_score: float = 0.6  # Minimum autocorrelation peak
    min_amplitude: float = 10.0  # Minimum peak-to-trough % CPU
    min_period: int = 2  # Minimum period in samples (Nyquist limit)
    max_period: int = 30  # Maximum period in samples


@dataclass
class OscillationResult:
    """Results of oscillation analysis (mirrors Go OscillationResult)."""

    detected: bool = False
    periodicity_score: float = 0.0  # Peak autocorrelation value (0.0-1.0)
    period: float = 0.0  # Detected period in seconds
    frequency: float = 0.0  # Cycles per second (Hz = 1/Period)
    amplitude: float = 0.0  # Peak-to-trough percentage
    stddev: float = 0.0  # Standard deviation of samples


@dataclass
class ContainerTimeseries:
    """CPU usage timeseries for a single container."""

    container_id: str
    container_short: str
    timestamps: list = field(default_factory=list)
    cpu_percent: list = field(default_factory=list)
    pod_uid: str | None = None
    qos_class: str | None = None


def autocorrelation(samples: np.ndarray, mean: float, variance: float, lag: int) -> float:
    """
    Compute normalized autocorrelation at a given lag.

    Returns a value in [-1, 1] where:
    - 1.0 means perfect positive correlation (signal repeats exactly)
    - 0.0 means no correlation (random noise)
    - -1.0 means perfect negative correlation (signal inverts)

    Matches Go implementation in detector.go.
    """
    if variance == 0 or lag >= len(samples):
        return 0.0

    n = len(samples)
    count = n - lag

    if count == 0:
        return 0.0

    # Sum of (x[i] - mean) * (x[i+lag] - mean)
    sum_val = np.sum((samples[:count] - mean) * (samples[lag:] - mean))

    # Normalize by variance to get correlation coefficient in [-1, 1]
    return sum_val / (count * variance)


def analyze_oscillation(samples: np.ndarray, config: OscillationConfig) -> OscillationResult:
    """
    Perform autocorrelation-based oscillation detection on a sample window.

    Mirrors the Analyze() method in detector.go.
    """
    result = OscillationResult()

    if len(samples) < config.window_size:
        return result

    # Use the most recent window_size samples
    window = samples[-config.window_size :]

    # Compute statistics
    mean = np.mean(window)
    variance = np.var(window)
    result.stddev = np.sqrt(variance)

    # Calculate amplitude (peak-to-trough)
    result.amplitude = float(np.max(window) - np.min(window))

    # Early exit if amplitude is below threshold
    if config.min_amplitude > 0 and result.amplitude < config.min_amplitude:
        return result

    # Compute autocorrelation for lags in [min_period, max_period]
    # Find the lag with the highest autocorrelation (strongest periodicity)
    best_lag = 0
    best_corr = 0.0

    for lag in range(config.min_period, config.max_period + 1):
        corr = autocorrelation(window, mean, variance, lag)
        if corr > best_corr:
            best_corr = corr
            best_lag = lag

    result.periodicity_score = best_corr

    # Convert lag (in samples) to period (in seconds, assuming 1Hz sampling)
    if best_lag > 0:
        result.period = float(best_lag)  # 1 sample = 1 second
        result.frequency = 1.0 / result.period

    # Detection criteria (all must be met):
    # 1. Periodicity score exceeds threshold
    # 2. Amplitude exceeds minimum threshold
    meets_periodicity = best_corr >= config.min_periodicity_score
    meets_amplitude = config.min_amplitude == 0 or result.amplitude >= config.min_amplitude

    if meets_periodicity and meets_amplitude:
        result.detected = True

    return result


def extract_label(labels: list[tuple], key: str) -> str | None:
    """Extract a label value from the labels list."""
    for k, v in labels:
        if k == key:
            return v
    return None


def load_and_prepare_data(filepath: Path) -> pd.DataFrame:
    """Load parquet file and extract labels into columns."""
    table = pq.read_table(filepath)
    df = table.to_pandas()

    df["container_id"] = df["labels"].apply(lambda x: extract_label(x, "container_id"))
    df["pod_uid"] = df["labels"].apply(lambda x: extract_label(x, "pod_uid"))
    df["qos_class"] = df["labels"].apply(lambda x: extract_label(x, "qos_class"))

    df["container_short"] = df["container_id"].apply(lambda x: x[:12] if x else "unknown")
    df["value"] = df["value_float"].combine_first(df["value_int"])

    return df


def compute_cpu_timeseries(df: pd.DataFrame) -> dict[str, ContainerTimeseries]:
    """
    Compute CPU usage timeseries for each container.

    Converts cumulative cpu.stat.usage_usec to CPU percentage.
    """
    # Filter to CPU usage metric
    cpu_metric = "cgroup.v2.cpu.stat.usage_usec"
    cpu_df = df[df["metric_name"] == cpu_metric].copy()

    if cpu_df.empty:
        print(f"Warning: No {cpu_metric} data found", file=sys.stderr)
        return {}

    # Sort by time within each container
    cpu_df = cpu_df.sort_values(["container_id", "time"])

    # Compute deltas
    cpu_df["value_delta"] = cpu_df.groupby("container_id")["value"].diff()
    cpu_df["time_delta_s"] = cpu_df.groupby("container_id")["time"].diff().dt.total_seconds()

    # Drop first row of each group and invalid deltas
    cpu_df = cpu_df.dropna(subset=["value_delta", "time_delta_s"])
    cpu_df = cpu_df[(cpu_df["value_delta"] >= 0) & (cpu_df["time_delta_s"] > 0)]

    # Compute CPU percentage (usec used / elapsed usec * 100)
    # 1 second of 1 core = 1,000,000 usec, so usec/sec / 10000 = % of one core
    cpu_df["cpu_percent"] = (cpu_df["value_delta"] / cpu_df["time_delta_s"]) / 10000

    # Build timeseries for each container
    timeseries = {}
    for container_id, group in cpu_df.groupby("container_id"):
        if pd.isna(container_id):
            continue

        ts = ContainerTimeseries(
            container_id=container_id,
            container_short=container_id[:12] if container_id else "unknown",
            timestamps=group["time"].tolist(),
            cpu_percent=group["cpu_percent"].tolist(),
            pod_uid=group["pod_uid"].iloc[0] if "pod_uid" in group.columns else None,
            qos_class=group["qos_class"].iloc[0] if "qos_class" in group.columns else None,
        )
        timeseries[container_id] = ts

    return timeseries


def analyze_containers(
    timeseries: dict[str, ContainerTimeseries],
    config: OscillationConfig,
) -> list[tuple[ContainerTimeseries, OscillationResult]]:
    """Analyze all containers for oscillation patterns."""
    results = []

    for container_id, ts in timeseries.items():
        samples = np.array(ts.cpu_percent)

        if len(samples) < config.window_size:
            # Not enough data
            result = OscillationResult()
            result.amplitude = float(np.max(samples) - np.min(samples)) if len(samples) > 1 else 0.0
            result.stddev = float(np.std(samples)) if len(samples) > 1 else 0.0
        else:
            result = analyze_oscillation(samples, config)

        results.append((ts, result))

    # Sort by periodicity score descending
    results.sort(key=lambda x: x[1].periodicity_score, reverse=True)

    return results


def print_results(
    results: list[tuple[ContainerTimeseries, OscillationResult]],
    config: OscillationConfig,
) -> None:
    """Print analysis results to console."""
    print("\n" + "=" * 80)
    print("CPU OSCILLATION DETECTION RESULTS")
    print("=" * 80)
    print("\nConfiguration:")
    print(f"  Window size:          {config.window_size} samples")
    print(f"  Min periodicity:      {config.min_periodicity_score}")
    print(f"  Min amplitude:        {config.min_amplitude}%")
    print(f"  Period range:         {config.min_period}-{config.max_period} seconds")

    # Separate detected vs not detected
    detected = [(ts, r) for ts, r in results if r.detected]
    not_detected = [(ts, r) for ts, r in results if not r.detected]

    if detected:
        print(f"\n{'OSCILLATION DETECTED':=^80}")
        print(f"\n{'Container':<14} {'Period':>8} {'Freq':>8} {'Score':>8} {'Amp':>8} {'StdDev':>8} {'QoS':<12}")
        print("-" * 80)
        for ts, result in detected:
            print(
                f"{ts.container_short:<14} "
                f"{result.period:>7.1f}s "
                f"{result.frequency:>7.3f}Hz "
                f"{result.periodicity_score:>7.2f} "
                f"{result.amplitude:>7.1f}% "
                f"{result.stddev:>7.2f} "
                f"{ts.qos_class or 'unknown':<12}"
            )
    else:
        print("\nNo oscillation patterns detected above threshold.")

    print(f"\n{'ALL CONTAINERS':=^80}")
    print(f"\n{'Container':<14} {'Samples':>8} {'Score':>8} {'Period':>8} {'Amp':>8} {'StdDev':>8} {'Detected':<8}")
    print("-" * 80)
    for ts, result in results[:20]:  # Top 20 by periodicity score
        samples = len(ts.cpu_percent)
        detected_str = "YES" if result.detected else "-"
        period_str = f"{result.period:.1f}s" if result.period > 0 else "-"
        print(
            f"{ts.container_short:<14} "
            f"{samples:>8} "
            f"{result.periodicity_score:>7.2f} "
            f"{period_str:>8} "
            f"{result.amplitude:>7.1f}% "
            f"{result.stddev:>7.2f} "
            f"{detected_str:<8}"
        )

    if len(results) > 20:
        print(f"\n... and {len(results) - 20} more containers")


def plot_autocorrelation(
    ts: ContainerTimeseries,
    config: OscillationConfig,
) -> go.Figure:
    """Create autocorrelation plot for a single container."""
    samples = np.array(ts.cpu_percent)

    if len(samples) < config.window_size:
        window = samples
    else:
        window = samples[-config.window_size :]

    mean = np.mean(window)
    variance = np.var(window)

    # Compute autocorrelation for all lags
    lags = list(range(1, min(len(window) // 2, config.max_period + 5)))
    correlations = [autocorrelation(window, mean, variance, lag) for lag in lags]

    fig = make_subplots(
        rows=2,
        cols=1,
        subplot_titles=(f"CPU Usage - {ts.container_short}", "Autocorrelation"),
        vertical_spacing=0.15,
    )

    # CPU timeseries
    fig.add_trace(
        go.Scatter(
            x=list(range(len(window))),
            y=window,
            mode="lines",
            name="CPU %",
            line={"color": "blue"},
        ),
        row=1,
        col=1,
    )

    # Autocorrelation
    colors = ["green" if config.min_period <= lag <= config.max_period else "gray" for lag in lags]
    fig.add_trace(
        go.Bar(
            x=lags,
            y=correlations,
            name="Autocorrelation",
            marker_color=colors,
        ),
        row=2,
        col=1,
    )

    # Add threshold line
    fig.add_hline(
        y=config.min_periodicity_score,
        line_dash="dash",
        line_color="red",
        annotation_text=f"Threshold ({config.min_periodicity_score})",
        row=2,
        col=1,
    )

    fig.update_xaxes(title_text="Sample Index", row=1, col=1)
    fig.update_xaxes(title_text="Lag (seconds)", row=2, col=1)
    fig.update_yaxes(title_text="CPU %", row=1, col=1)
    fig.update_yaxes(title_text="Correlation", row=2, col=1)

    fig.update_layout(
        title=f"Oscillation Analysis: {ts.container_short}",
        height=600,
        showlegend=False,
    )

    return fig


def plot_all_results(
    results: list[tuple[ContainerTimeseries, OscillationResult]],
    config: OscillationConfig,
    output_path: Path | None = None,
    top_n: int = 6,
) -> None:
    """Create a combined plot showing top containers by periodicity score."""
    # Take top N containers
    top_results = results[:top_n]

    fig = make_subplots(
        rows=top_n,
        cols=2,
        subplot_titles=[
            title
            for ts, r in top_results
            for title in (
                f"{ts.container_short} (score={r.periodicity_score:.2f})",
                f"Autocorrelation (period={r.period:.1f}s)",
            )
        ],
        vertical_spacing=0.08,
        horizontal_spacing=0.1,
    )

    for idx, (ts, result) in enumerate(top_results, 1):
        samples = np.array(ts.cpu_percent)

        if len(samples) < config.window_size:
            window = samples
        else:
            window = samples[-config.window_size :]

        mean = np.mean(window)
        variance = np.var(window)

        # CPU timeseries
        color = "red" if result.detected else "blue"
        fig.add_trace(
            go.Scatter(
                x=list(range(len(window))),
                y=window,
                mode="lines",
                name=ts.container_short,
                line={"color": color},
                showlegend=False,
            ),
            row=idx,
            col=1,
        )

        # Autocorrelation
        lags = list(range(1, min(len(window) // 2, config.max_period + 5)))
        correlations = [autocorrelation(window, mean, variance, lag) for lag in lags]

        bar_colors = ["green" if config.min_period <= lag <= config.max_period else "lightgray" for lag in lags]
        fig.add_trace(
            go.Bar(
                x=lags,
                y=correlations,
                marker_color=bar_colors,
                showlegend=False,
            ),
            row=idx,
            col=2,
        )

        # Threshold line
        fig.add_hline(
            y=config.min_periodicity_score,
            line_dash="dash",
            line_color="red",
            row=idx,
            col=2,
        )

    fig.update_layout(
        title="CPU Oscillation Detection - Top Containers by Periodicity Score",
        height=200 * top_n,
    )

    if output_path:
        if output_path.suffix == ".html":
            fig.write_html(output_path)
        else:
            fig.write_image(output_path)
        print(f"Saved plot to {output_path}")
    else:
        fig.show()


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Detect CPU oscillation patterns using autocorrelation",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
    # Basic analysis
    uv run scripts/oscillation_detector.py metrics.parquet

    # Lower thresholds for more sensitive detection
    uv run scripts/oscillation_detector.py metrics.parquet --threshold 0.4 --min-amplitude 5.0

    # Generate HTML report with plots
    uv run scripts/oscillation_detector.py metrics.parquet --plot --output report.html

    # Analyze specific container
    uv run scripts/oscillation_detector.py metrics.parquet --container abc123456789
        """,
    )
    parser.add_argument("input", type=Path, help="Input parquet file")
    parser.add_argument(
        "-t",
        "--threshold",
        type=float,
        default=0.6,
        help="Minimum periodicity score (autocorrelation threshold, default: 0.6)",
    )
    parser.add_argument(
        "-a",
        "--min-amplitude",
        type=float,
        default=10.0,
        help="Minimum amplitude (peak-to-trough %%, default: 10.0)",
    )
    parser.add_argument(
        "-w",
        "--window",
        type=int,
        default=60,
        help="Analysis window size in samples (default: 60)",
    )
    parser.add_argument(
        "--min-period",
        type=int,
        default=2,
        help="Minimum period to detect in seconds (default: 2)",
    )
    parser.add_argument(
        "--max-period",
        type=int,
        default=30,
        help="Maximum period to detect in seconds (default: 30)",
    )
    parser.add_argument(
        "-c",
        "--container",
        type=str,
        help="Analyze only this container (prefix match)",
    )
    parser.add_argument(
        "-p",
        "--plot",
        action="store_true",
        help="Generate visualization plots",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=Path,
        help="Output file for plots (HTML or image)",
    )
    parser.add_argument(
        "-n",
        "--top",
        type=int,
        default=6,
        help="Number of containers to show in plot (default: 6)",
    )

    args = parser.parse_args()

    if not args.input.exists():
        print(f"Error: File not found: {args.input}", file=sys.stderr)
        sys.exit(1)

    config = OscillationConfig(
        window_size=args.window,
        min_periodicity_score=args.threshold,
        min_amplitude=args.min_amplitude,
        min_period=args.min_period,
        max_period=args.max_period,
    )

    print(f"Loading {args.input}...")
    df = load_and_prepare_data(args.input)
    print(f"Loaded {len(df):,} rows")

    print("Computing CPU timeseries...")
    timeseries = compute_cpu_timeseries(df)
    print(f"Found {len(timeseries)} containers")

    if args.container:
        # Filter to specific container
        filtered = {
            k: v
            for k, v in timeseries.items()
            if k.startswith(args.container) or v.container_short.startswith(args.container)
        }
        if not filtered:
            print(f"Error: No container matching '{args.container}'", file=sys.stderr)
            sys.exit(1)
        timeseries = filtered
        print(f"Filtered to {len(timeseries)} container(s)")

    print("Analyzing for oscillation patterns...")
    results = analyze_containers(timeseries, config)

    print_results(results, config)

    if args.plot or args.output:
        plot_all_results(results, config, args.output, args.top)


if __name__ == "__main__":
    main()

File: q_branch/fine-grained-monitor/scripts/oscillation_detector.rs
#!/usr/bin/env -S cargo +nightly -Zscript

---
[dependencies]
parquet = { version = "54", features = ["arrow"] }
arrow = "54"
clap = { version = "4", features = ["derive"] }
anyhow = "1"

[profile.dev]
opt-level = 3
---

//! Detect CPU oscillation patterns in container metrics using autocorrelation.
//!
//! Rust port of oscillation_detector.py for faster processing of large datasets.
//!
//! Usage:
//!     ./oscillation_detector.rs metrics.parquet
//!     ./oscillation_detector.rs metrics.parquet --threshold 0.3 --min-amplitude 5.0

use anyhow::{Context, Result};
use arrow::array::{Array, Float64Array, MapArray, StringArray, StructArray, UInt64Array};
use arrow::datatypes::DataType;
use clap::Parser;
use parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder;
use std::collections::HashMap;
use std::fs::File;
use std::path::PathBuf;

#[derive(Parser, Debug)]
#[command(name = "oscillation_detector")]
#[command(about = "Detect CPU oscillation patterns using autocorrelation")]
struct Args {
    /// Input parquet file
    input: PathBuf,

    /// Minimum periodicity score (autocorrelation threshold)
    #[arg(short = 't', long, default_value = "0.6")]
    threshold: f64,

    /// Minimum amplitude (peak-to-trough %)
    #[arg(short = 'a', long, default_value = "10.0")]
    min_amplitude: f64,

    /// Analysis window size in samples
    #[arg(short = 'w', long, default_value = "60")]
    window: usize,

    /// Minimum period to detect in seconds
    #[arg(long, default_value = "2")]
    min_period: usize,

    /// Maximum period to detect in seconds
    #[arg(long, default_value = "30")]
    max_period: usize,

    /// Analyze only this container (prefix match)
    #[arg(short = 'c', long)]
    container: Option<String>,

    /// Show top N containers
    #[arg(short = 'n', long, default_value = "20")]
    top: usize,
}

#[derive(Debug, Clone)]
struct OscillationConfig {
    window_size: usize,
    min_periodicity_score: f64,
    min_amplitude: f64,
    min_period: usize,
    max_period: usize,
}

#[derive(Debug, Clone, Default)]
struct OscillationResult {
    detected: bool,
    periodicity_score: f64,
    period: f64,
    frequency: f64,
    amplitude: f64,
    stddev: f64,
}

#[derive(Debug)]
struct ContainerTimeseries {
    container_id: String,
    container_short: String,
    #[allow(dead_code)]
    timestamps: Vec<i64>,
    cpu_percent: Vec<f64>,
    qos_class: Option<String>,
}

/// Compute normalized autocorrelation at a given lag
fn autocorrelation(samples: &[f64], mean: f64, variance: f64, lag: usize) -> f64 {
    if variance == 0.0 || lag >= samples.len() {
        return 0.0;
    }

    let n = samples.len();
    let count = n - lag;
    if count == 0 {
        return 0.0;
    }

    let sum: f64 = (0..count)
        .map(|i| (samples[i] - mean) * (samples[i + lag] - mean))
        .sum();

    sum / (count as f64 * variance)
}

/// Analyze samples for oscillation patterns
fn analyze_oscillation(samples: &[f64], config: &OscillationConfig) -> OscillationResult {
    let mut result = OscillationResult::default();

    if samples.len() < config.window_size {
        if samples.len() > 1 {
            let min = samples.iter().cloned().fold(f64::INFINITY, f64::min);
            let max = samples.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
            result.amplitude = max - min;

            let mean: f64 = samples.iter().sum::<f64>() / samples.len() as f64;
            let variance: f64 =
                samples.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / samples.len() as f64;
            result.stddev = variance.sqrt();
        }
        return result;
    }

    // Use the most recent window_size samples
    let window = &samples[samples.len() - config.window_size..];

    // Compute statistics
    let mean: f64 = window.iter().sum::<f64>() / window.len() as f64;
    let variance: f64 =
        window.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / window.len() as f64;
    result.stddev = variance.sqrt();

    // Calculate amplitude
    let min = window.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = window.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    result.amplitude = max - min;

    // Early exit if amplitude below threshold
    if config.min_amplitude > 0.0 && result.amplitude < config.min_amplitude {
        return result;
    }

    // Find best autocorrelation lag
    let mut best_lag = 0;
    let mut best_corr = 0.0;

    for lag in config.min_period..=config.max_period {
        let corr = autocorrelation(window, mean, variance, lag);
        if corr > best_corr {
            best_corr = corr;
            best_lag = lag;
        }
    }

    result.periodicity_score = best_corr;

    if best_lag > 0 {
        result.period = best_lag as f64;
        result.frequency = 1.0 / result.period;
    }

    // Detection criteria
    let meets_periodicity = best_corr >= config.min_periodicity_score;
    let meets_amplitude = config.min_amplitude == 0.0 || result.amplitude >= config.min_amplitude;

    if meets_periodicity && meets_amplitude {
        result.detected = true;
    }

    result
}

/// Extract a label value from the labels map
fn extract_label(labels: &[(String, String)], key: &str) -> Option<String> {
    labels
        .iter()
        .find(|(k, _)| k == key)
        .map(|(_, v)| v.clone())
}

fn main() -> Result<()> {
    let args = Args::parse();

    if !args.input.exists() {
        anyhow::bail!("File not found: {:?}", args.input);
    }

    let config = OscillationConfig {
        window_size: args.window,
        min_periodicity_score: args.threshold,
        min_amplitude: args.min_amplitude,
        min_period: args.min_period,
        max_period: args.max_period,
    };

    eprintln!("Loading {:?}...", args.input);

    let file = File::open(&args.input).context("Failed to open file")?;
    let builder = ParquetRecordBatchReaderBuilder::try_new(file)?;

    // Project only needed columns for better performance
    let schema = builder.schema();
    let projection_indices: Vec<usize> = ["metric_name", "time", "value_int", "value_float", "labels"]
        .iter()
        .filter_map(|name| schema.index_of(name).ok())
        .collect();

    // Build the projection mask while we still have access to parquet_schema
    let projection_mask = parquet::arrow::ProjectionMask::roots(
        builder.parquet_schema(),
        projection_indices,
    );

    let reader = builder
        .with_projection(projection_mask)
        .with_batch_size(65536)  // Larger batches for efficiency
        .build()?;

    // Collect CPU usage data per container
    // Key: container_id, Value: (timestamps, cpu_values, qos_class, last_value, last_time)
    let mut container_data: HashMap<String, (Vec<i64>, Vec<f64>, Option<String>, f64, i64)> =
        HashMap::new();

    let mut total_rows = 0u64;
    let cpu_metric = "cgroup.v2.cpu.stat.usage_usec";

    for batch_result in reader {
        let batch = batch_result?;
        total_rows += batch.num_rows() as u64;

        // Get columns
        let metric_names = batch
            .column_by_name("metric_name")
            .and_then(|c| c.as_any().downcast_ref::<StringArray>())
            .context("Missing metric_name column")?;

        let times = batch
            .column_by_name("time")
            .context("Missing time column")?;

        let values_int = batch
            .column_by_name("value_int")
            .and_then(|c| c.as_any().downcast_ref::<UInt64Array>());

        let values_float = batch
            .column_by_name("value_float")
            .and_then(|c| c.as_any().downcast_ref::<Float64Array>());

        let labels_col = batch
            .column_by_name("labels")
            .context("Missing labels column")?;

        // Extract timestamp values
        let time_values: Vec<i64> = match times.data_type() {
            DataType::Timestamp(_, _) => {
                let ts_array = times
                    .as_any()
                    .downcast_ref::<arrow::array::TimestampMillisecondArray>()
                    .context("Failed to cast timestamp")?;
                (0..ts_array.len()).map(|i| ts_array.value(i)).collect()
            }
            _ => anyhow::bail!("Unexpected time column type"),
        };

        // Process each row
        for row in 0..batch.num_rows() {
            let metric = metric_names.value(row);
            if metric != cpu_metric {
                continue;
            }

            // Get value
            let value = if let Some(arr) = values_float {
                if !arr.is_null(row) {
                    arr.value(row)
                } else if let Some(int_arr) = values_int {
                    if !int_arr.is_null(row) {
                        int_arr.value(row) as f64
                    } else {
                        continue;
                    }
                } else {
                    continue;
                }
            } else if let Some(int_arr) = values_int {
                if !int_arr.is_null(row) {
                    int_arr.value(row) as f64
                } else {
                    continue;
                }
            } else {
                continue;
            };

            let time = time_values[row];

            // Extract labels
            let labels = extract_labels_from_column(labels_col, row)?;
            let container_id = match extract_label(&labels, "container_id") {
                Some(id) => id,
                None => continue,
            };
            let qos_class = extract_label(&labels, "qos_class");

            // Update container data
            let entry = container_data
                .entry(container_id.clone())
                .or_insert_with(|| (Vec::new(), Vec::new(), qos_class.clone(), -1.0, 0));

            // Compute delta if we have a previous value
            if entry.3 >= 0.0 && time > entry.4 {
                let value_delta = value - entry.3;
                let time_delta_ms = time - entry.4;

                if value_delta >= 0.0 && time_delta_ms > 0 {
                    // Convert usec delta to CPU percentage
                    // usec/ms = usec / (ms * 1000) * 100 = usec / ms / 10
                    let cpu_percent = value_delta / (time_delta_ms as f64) / 10.0;
                    entry.0.push(time);
                    entry.1.push(cpu_percent);
                }
            }

            entry.3 = value;
            entry.4 = time;
            if entry.2.is_none() {
                entry.2 = qos_class;
            }
        }
    }

    eprintln!("Loaded {} rows", total_rows);

    // Build timeseries
    let mut timeseries: Vec<ContainerTimeseries> = container_data
        .into_iter()
        .map(|(id, (timestamps, cpu_percent, qos_class, _, _))| {
            let container_short = if id.len() > 12 {
                id[..12].to_string()
            } else {
                id.clone()
            };
            ContainerTimeseries {
                container_id: id,
                container_short,
                timestamps,
                cpu_percent,
                qos_class,
            }
        })
        .collect();

    eprintln!("Found {} containers", timeseries.len());

    // Filter by container if specified
    if let Some(ref prefix) = args.container {
        timeseries.retain(|ts| {
            ts.container_id.starts_with(prefix) || ts.container_short.starts_with(prefix)
        });
        eprintln!("Filtered to {} container(s)", timeseries.len());
    }

    // Analyze each container
    let mut results: Vec<(ContainerTimeseries, OscillationResult)> = timeseries
        .into_iter()
        .map(|ts| {
            let result = analyze_oscillation(&ts.cpu_percent, &config);
            (ts, result)
        })
        .collect();

    // Sort by periodicity score descending
    results.sort_by(|a, b| {
        b.1.periodicity_score
            .partial_cmp(&a.1.periodicity_score)
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    // Print results
    println!();
    println!("{}", "=".repeat(80));
    println!("CPU OSCILLATION DETECTION RESULTS");
    println!("{}", "=".repeat(80));
    println!();
    println!("Configuration:");
    println!("  Window size:          {} samples", config.window_size);
    println!("  Min periodicity:      {}", config.min_periodicity_score);
    println!("  Min amplitude:        {}%", config.min_amplitude);
    println!(
        "  Period range:         {}-{} seconds",
        config.min_period, config.max_period
    );

    // Detected containers
    let detected: Vec<_> = results.iter().filter(|(_, r)| r.detected).collect();

    if !detected.is_empty() {
        println!();
        println!("{:=^80}", "OSCILLATION DETECTED");
        println!();
        println!(
            "{:<14} {:>8} {:>8} {:>8} {:>8} {:>8} {:<12}",
            "Container", "Period", "Freq", "Score", "Amp", "StdDev", "QoS"
        );
        println!("{}", "-".repeat(80));

        for (ts, result) in &detected {
            println!(
                "{:<14} {:>7.1}s {:>7.3}Hz {:>7.2} {:>7.1}% {:>7.2} {:<12}",
                ts.container_short,
                result.period,
                result.frequency,
                result.periodicity_score,
                result.amplitude,
                result.stddev,
                ts.qos_class.as_deref().unwrap_or("unknown")
            );
        }
    } else {
        println!();
        println!("No oscillation patterns detected above threshold.");
    }

    // All containers
    println!();
    println!("{:=^80}", "ALL CONTAINERS");
    println!();
    println!(
        "{:<14} {:>8} {:>8} {:>8} {:>8} {:>8} {:<8}",
        "Container", "Samples", "Score", "Period", "Amp", "StdDev", "Detected"
    );
    println!("{}", "-".repeat(80));

    for (ts, result) in results.iter().take(args.top) {
        let detected_str = if result.detected { "YES" } else { "-" };
        let period_str = if result.period > 0.0 {
            format!("{:.1}s", result.period)
        } else {
            "-".to_string()
        };

        println!(
            "{:<14} {:>8} {:>7.2} {:>8} {:>7.1}% {:>7.2} {:<8}",
            ts.container_short,
            ts.cpu_percent.len(),
            result.periodicity_score,
            period_str,
            result.amplitude,
            result.stddev,
            detected_str
        );
    }

    if results.len() > args.top {
        println!();
        println!("... and {} more containers", results.len() - args.top);
    }

    Ok(())
}

/// Extract labels from the labels column for a specific row
fn extract_labels_from_column(col: &dyn Array, row: usize) -> Result<Vec<(String, String)>> {
    // The labels column is a Map<String, String>
    let map_array = col
        .as_any()
        .downcast_ref::<MapArray>()
        .context("Labels column is not a MapArray")?;

    if map_array.is_null(row) {
        return Ok(vec![]);
    }

    let start = map_array.value_offsets()[row] as usize;
    let end = map_array.value_offsets()[row + 1] as usize;

    let entries = map_array.entries();
    let struct_array = entries
        .as_any()
        .downcast_ref::<StructArray>()
        .context("Map entries is not a StructArray")?;

    let keys = struct_array
        .column(0)
        .as_any()
        .downcast_ref::<StringArray>()
        .context("Missing key column in labels")?;

    let vals = struct_array
        .column(1)
        .as_any()
        .downcast_ref::<StringArray>()
        .context("Missing value column in labels")?;

    let mut result = Vec::with_capacity(end - start);
    for i in start..end {
        if !keys.is_null(i) && !vals.is_null(i) {
            result.push((keys.value(i).to_string(), vals.value(i).to_string()));
        }
    }

    Ok(result)
}
